peps/pep-0694.rst

```diff
@@ -283,639 +383,568 @@ the following keys:
-
-
 .. _staged-preview:
 
 Stage Previews
-~~~~~~~~~~~~~~
+--------------
 
 The ability to preview staged releases before they are published is an important feature of this
 PEP, enabling an additional level of last-mile testing before the release is available to the
 public.  Indexes **MAY** provide this functionality through the URL provided in the ``stage``
-sub-key of the :ref:`links key <session-links>` returned when the session is created.  The ``stage``
-URL can be passed to installers such as ``pip`` by setting the `--extra-index-url
+sub-key of the :ref:`links key <publishing-session-links>` returned when
+the Publishing Session is created.
+The ``stage`` URL can be passed to installers such as ``pip`` by setting the `--extra-index-url
 <https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-extra-index-url>`__ flag to this value.
 Multiple stages can even be previewed by repeating this flag with multiple values.
 
-In the future, it may be valuable to include something like a ``Stage-Token`` header to the `Simple
-Repository API <https://packaging.python.org/en/latest/specifications/simple-repository-api/>`_
-requests or the :pep:`691` JSON-based Simple API, with the value from the ``session-token`` sub-key
-of the JSON response to the session creation request.  Multiple ``Stage-Token`` headers could be
-allowed, and installers could support enabling stage previews by adding a ``--staged <token>`` or
-similarly named option to set the ``Stage-Token`` header at the command line.  This feature is not
-currently support, nor proposed by this PEP, though it could be proposed by a separate PEP in the
-future.
-
 In either case, the index will return views that expose the staged releases to the installer tool,
 making them available to download and install into virtual environments built for that last-mile
 testing.  The former option allows for existing installers to preview staged releases with no
 changes, although perhaps in a less user-friendly way.  The latter option can be a better user
 experience, but the details of this are left to installer tool maintainers.
 
 
-.. _session-errors:
+.. _file-upload-mechanisms:
 
-Errors
-------
+File Upload Mechanisms
+----------------------
 
-All error responses that contain content will have a body that looks like:
+Servers **MUST** implement :ref:`required file upload mechanisms <required-file-upload-mechanisms>`.
+Such mechanisms serve as a fallback if no server specific implementations exist.
 
-.. code-block:: json
+Each major version of the Upload API **MUST** specify at least one required File Upload Mechanism.
 
-    {
-      "meta": {
-        "api-version": "2.0"
-      },
-      "message": "...",
-      "errors": [
-        {
-          "source": "...",
-          "message": "..."
-        }
-      ]
-    }
+New required mechanisms **MUST NOT** be added
+and existing required mechanisms **MUST NOT** be removed
+without an update to the :ref:`major version <versioning>`.
```

Comment:
```suggestion
without an update to the :ref:`major version <versioning>`.  Any server-specific or experimental mechanisms added or removed **MUST NOT** change the major version number of this specification.
```

Reply:
Should we also limit the minor version number?  E.g. what if `myindex.example.org` adds a `"wishful-thinking"` mechanism?  Should that index be allowed to call their version number `"2.1"`?

Reply:
This is specifically about required mechanisms. Those should only be changed on major version and via PEP.

Reply:
```suggestion
without an update to the :ref:`major version <versioning>`.  Any server-specific or experimental mechanisms added or removed **MUST NOT** change the major or minor version number of this specification.
```